cql3/type_json.cc

```diff
@@ -78,8 +78,12 @@ static int64_t to_int64_t(const rjson::value& value) {
         return value.GetInt();
     } else if (value.IsUint()) {
         return value.GetUint();
-    } else if (value.GetUint64()) {
+    } else if (value.IsUint64()) {
         return value.GetUint64(); //NOTICE: large uint64_t values will get overflown
+    } else if (value.IsDouble()) {
+        return value.GetDouble();
```

Comment:
I agree. What happens with DynamoDB and MongoDB? They are schemaless, but if you declare something as a sort key, it suddenly has a type. Although I see DynamoDB has just a number type, not separate integer and float types.

Reply:
Right - DynamoDB indeed, like JSON, has just one "number" type so there is no separate notions of integer and floating point. But Scylla does have types, and the above assignment will assign 2, not 2.5, into the column, which sounds to me like a bad idea, and unlikely to be something that a real user will want to do deliberately - and if it's done by mistake an error will be a better result.

Reply:
Well, I'm trying to go for "right", not just bug compatibility with older versions. Especially since only one user complained that this changed - and they only cared about integers in scientific notation - not about floating point like 2.5 working.

Your point about 2.000000001 is a very good point. Not because this example (which is clearly NOT a whole number) and not because of any 32-bit integer - the 32-bit integers can all be accurately represented in a double because double has more digits. However, we would have a bug with very high bigint (64-bit integers): The highest bigint is 2^63 - 1 = 9223372036854775807. If this is ever saved as a double, the last digit of precision will not fit and will not be reproduced correctly  :-( This means that if we write for example 9223372036854775.807E3 and RapidJson parses it as a double, when we save it into a bigint we can loose one digit of precision. This is a bug. We'll need to remember (e.g., in the form of a failing test) that we have this bug. Maybe we even need to recognize this case and error-out  (double-to-integer conversion when the double is larger than some threshold) so the user doesn't loose accuracy.

But in more "reasonable" cases, like the one asked for the original user (with numbers around 1E9), this will be 100% accurate.

To be 100% accurate in every case we'll need to fix RapidJson to use unlimited precision when parsing numbers. This is what we do in Alternator. But it's not how RapidJson works...